client/telemetry/src/layer.rs

```diff
@@ -0,0 +1,146 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use futures::channel::mpsc;
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::sync::Arc;
+use tracing::{Event, Id, Subscriber};
+use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
+
+/// Span name used to report the telemetry.
+pub const TELEMETRY_LOG_SPAN: &str = "telemetry-logger";
+
+/// `Layer` that handles the logs for telemetries.
+#[derive(Debug)]
+pub struct TelemetryLayer(Senders);
+
+impl TelemetryLayer {
+	/// Create a new [`TelemetryLayer`] using the [`Senders`] provided in argument.
+	pub fn new(senders: Senders) -> Self {
+		Self(senders)
+	}
+}
+
+impl<S> Layer<S> for TelemetryLayer
+where
+	S: Subscriber + for<'a> LookupSpan<'a>,
+{
+	fn on_event(&self, event: &Event<'_>, ctx: Context<S>) {
+		if event.metadata().target() != TELEMETRY_LOG_SPAN {
+			return;
+		}
+
+		if let Some(span) = ctx.lookup_current() {
+			let parents = span.parents();
+
+			if let Some(span) = std::iter::once(span)
+				.chain(parents)
+				.find(|x| x.name() == TELEMETRY_LOG_SPAN)
+			{
+				let id = span.id();
+				if let Some(sender) = (self.0).0.lock().get_mut(&id) {
+					let mut attrs = TelemetryAttrs::new(id);
+					let mut vis = TelemetryAttrsVisitor(&mut attrs);
+					event.record(&mut vis);
+
+					if let TelemetryAttrs {
+						verbosity: Some(verbosity),
+						json: Some(json),
+						..
+					} = attrs
+					{
+						let _ = sender.try_send((
+							verbosity
+								.try_into()
+								.expect("telemetry log message verbosity are u8; qed"),
+							json,
+						));
+					} else {
+						panic!(
+							"missing fields in telemetry log: {:?}. This can happen if \
+							`tracing::info_span!` is (mis-)used with the telemetry target \
+							directly; you should use the `telemetry!` macro.",
+							event,
+						);
+					}
+				}
+			}
+		}
+	}
+}
+
+#[derive(Debug)]
+struct TelemetryAttrs {
+	verbosity: Option<u64>,
+	json: Option<String>,
+	id: Id,
+}
+
+impl TelemetryAttrs {
+	fn new(id: Id) -> Self {
+		Self {
+			verbosity: None,
+			json: None,
+			id,
+		}
+	}
+}
+
+#[derive(Debug)]
+struct TelemetryAttrsVisitor<'a>(&'a mut TelemetryAttrs);
+
+impl<'a> tracing::field::Visit for TelemetryAttrsVisitor<'a> {
+	fn record_debug(&mut self, _field: &tracing::field::Field, _value: &dyn std::fmt::Debug) {
+		// noop
+	}
+
+	fn record_u64(&mut self, field: &tracing::field::Field, value: u64) {
+		if field.name() == "verbosity" {
+			(*self.0).verbosity = Some(value)
+		}
+	}
+
+	fn record_str(&mut self, field: &tracing::field::Field, value: &str) {
+		if field.name() == "json" {
+			// NOTE: this is a hack to inject the span id into the json
+			let mut message = format!(r#"{{"id":{},"#, (*self.0).id.into_u64());
```

Comment:
And where is this json object "closed" with the "}"?

Reply:
The value `json` already has the closing `}`. It's working, I have tested

Reply:
Where does this closing bracket is coming from?

Reply:
When `telemetry!` is used, the JSON is created. https://github.com/paritytech/substrate/blob/ca23c7a/client/telemetry/src/lib.rs#L317

Reply:
Are you saying that `value` can for example contain `message:"foo"}` ? This looks extremely sketchy to me.

Reply:
This is extremely sketchy.

Reply:
I added a test that checks for the first character:
```
			if value.chars().next() == Some('{') {
				eprintln!(
					"Invalid value for JSON in telemetry logging: \
					the attribute's value `json` doesn't start with the character `{{`",
				);
			}
```

But here are things you need to know:
1. it is the way it is partly because tracing can't use other things than primitives like strings and numbers in the attribute values (slog was able to use anything that is serializable)
3. because of this issue, I'm forced to serialize in JSON "early", meaning I have to serialize to json where the log is emitted, by the `telemetry!` macro. The json passed in attribute is always valid unless the log is emitted manually by the user. @bkchr this is where the closing bracket come from (I just discard the opening one) @tomaka the only way it could break is if the developer emits the log manually and don't serialize the json properly
4. for the telemetry to work now that we might have multiple node reaching out the same server, I need to inject an id of the node. I'm using the span id for that (as u64)
5. the span id is only accessible from within the layer, I couldn't find a way to get it in place of the log, but I'm asking the question now on discord to see if there is any other alternative
6. the logging doesn't work from within the layer, this is why I'm using `eprintln!` here as @bkchr suggested

One other alternative would be to deserialize the json, add the `id`, re-serialize the json.

Reply:
Weird idea :P 

```
{
   "id": "yolo",
   "wrapped": data that we got from the macro.
}
```

Reply:
At least it's less hacky than what I did ^_^'

Reply:
Done in ddb53d051 and PR https://github.com/paritytech/substrate-telemetry/pull/306